### ISSUE TITLE
Dramatically slow rebuild times against a single file touch

### ISSUE THREAD

**Issue by bucaran** [broccolijs/broccoli-funnel#27]
I am currently profiling / adding temporal instrumentation debug console logs to several broccoli plugins in my project in order to find the cause of horribly slow rebuild times (2minutes).

I am particularly looking at _20s_ to _35s_ seconds in between handling input trees (adding up to a 1m of the build process). The suspects are:

+ `Funnel.prototype.read`
+ `Funnel.prototype.handleReadTree`
+ `Funnel.prototype.cleanup`.

What I am trying to understand is why does Broccoli use up to 30s in average to process input trees when I touch just a single file in my app? :sos:

**Comment by rwjblue** [comment]
It seems somewhat unlikely that this is the result of an issue with broccoli-funnel (but definitely still possible).  Can you share the results of the slow trees logging?  What does your `Brocfile.js` look like? What are the input options to the `Funnel` instance that is slow?  How large are your input trees (both by file size and file count)?  

Most likely this is due to some trees improperly invalidating the tree (and not caching).  The recent refactor towards broccoli-plugin and the newer API's will allow a static dependency graph (therefore allowing us to prevent calling `.read` on all trees, rather only call `.build` on those trees whose inputs have changed).

**Comment by bucaran** [comment]
By inspecting console traces I ended up in Broc's `builder.js` where `Funnel.proto.read` is used several times.

I should have specified that this is an `ember-cli` app and my Brocfile is pretty basic:

```js
var EmberApp = require('ember-cli/lib/broccoli/ember-app');
var environment = EmberApp.env();

var emberCliConfig = {
  fingerprint: { enabled: false },
  storeConfigInMeta: false
};

var app = new EmberApp(emberCliConfig);

app.import(/*VENDOR INTERNATIONALIZATION JAVASCRIPTS'*/);

app.import(/* BOWER COMPONENTS (moment, jquery ui, antiscroll, fastclick, wanakana, font-awesome)*/)
app.import(/* SPECIFIC APP ICONS */)

```

**Comment by rwjblue** [comment]
> I should have specified that this is an ember-cli app and my Brocfile is pretty basic:

It may seem basic, but your Brocfile is quite complex (take a look at what [ember-cli/lib/broccoli/ember-app.js](https://github.com/ember-cli/ember-cli/blob/master/lib/broccoli/ember-app.js) is doing).
 
> > How large are your input trees (both by file size and file count)
> How can I determine this?

Check the size on disk of the directory that is used for the slow instance of broccoli-funnel, look at its size and number of files.

**Comment by bucaran** [comment]
The number of directories is relatively large, around 200 MB megabytes.

I commented those strings due to privacy reasons, but we are talking about no more than 5 bower components, including jQuery UI :confused: 

:new:  A new discovery is that the `tmp` dir on a mbp _without_ SSD is showing tons of unremoved `tree_merger_tmp_des...` directories that add up to 170 MB, while the same directory on a mbp _with_ SSD is correctly cleaned up to a few bytes.

**Comment by rwjblue** [comment]
You don't mention SSD at all in your description. SSD's will be much faster than HDD's for the same CPU/mem setup.

**Comment by rwjblue** [comment]
Also, you can safely `rm -rf tmp` to clear the temp folders if cleanup wasn't completed properly.

**Comment by bucaran** [comment]
My bad, but the difference is too large. We are talking about 4s VS 2~3 minutes.

Also, what is with these tree_merger_tmp_des directories not getting cleaned up on the slow mbps?

One more thing, even if I clean up tmp the builds don't improve more than a 10 percent.

**Comment by rwjblue** [comment]
> My bad, but the difference is too large. We are talking about 4s VS 2~3 minutes.

Every stage in a broccoli pipeline is writing to disk, in an ember-cli build there are likely at least 20-30 stages that have to be gone through.  That leads to massively different results between HDD and SDD (do to I/O throughput).

**Comment by bucaran** [comment]
Understood. 

Any hints regarding the tree_merger_tmp_des directories not getting cleaned up only on non SDD mbps?

**Comment by bucaran** [comment]
@raytiley You may be right, you are probably right, but do you hav any hints regarding the `tree_merger_tmp_dest` directories not getting cleaned up only on regular HDD mbp? 

Why is Broccoli only cleaning up on SDDs?

**Comment by stefanpenner** [comment]
@rwjblue fyi this is already being tracked on https://github.com/ember-cli/ember-cli/issues/4460

**Comment by bucaran** [comment]
@stefanpenner Thanks. I would appreciate if you could address:

> #### Why is Broccoli only cleaning up on SDDs?

**Comment by stefanpenner** [comment]
> @stefanpenner Thanks. I would appreciate if you could provide an answer or any hints regarding:
> 
> Why is Broccoli only cleaning up on SDDs?

if cleanup isn't happening it should be uniform regardless of device type. The only thing that could cause a difference would be differences in the underlying FS. Which are likely to not differ between SSD / HDD

**Comment by bucaran** [comment]
Both are last generation mbps with exactly the same configuration. The only difference is one has flash storage and the other regular HDDs. The fact of the matter remains, HDDs are not getting properly cleaned. Have you had this happen to you before? Any hints?

BTW I have added instrumentation to several modules to help me track this down, and concluded the bottlenecks occur in Funnels and Broc#builder. I hope that helps :disappointed_relieved: 

One more thing:

> Every stage in a broccoli pipeline is writing to disk, in an ember-cli build there are likely at least 20-30 stages that have to be gone through. That leads to massively different results between HDD and SDD (do to I/O throughput).

You as another core member may have your own opinion on the subject, or perhaps agree with @rwjblue? Do you think the only problem is due to SDD vs HDD? Or could there be bugs somewhere else?

Thanks for your time and sorry for creating so many issues lately!

**Comment by bucaran** [comment]
I don't think the problem is there, but have you ever considered replacing RSVP with bluebird for a probably faster promise implementation? Don't think that's going to happen anytime soon anyway.

**Comment by bucaran** [comment]
@stefanpenner Both are last generation mbps with exactly the same configuration. The only difference is one has flash storage and the other a regular HDD. The fact of the matter remains, the HDD ones (3 machines are presenting the same issue) are not getting properly cleaned. Have you had this happen to you before? Any hints?

BTW I have added instrumentation to several modules to help me track this down and positively concluded the bottlenecks occur in calls to `Funnel.proto.read` in `Broc#builder`. I hope that helps :disappointed_relieved: 

One more thing:

> Every stage in a broccoli pipeline is writing to disk, in an ember-cli build there are likely at least 20-30 stages that have to be gone through. That leads to massively different results between HDD and SDD (do to I/O throughput).

You as another core member may have your own opinion on the subject, or perhaps agree with @rwjblue? Do you think the only problem is due to SDD vs HDD? Or could there be bugs somewhere else?

Thanks for your time and sorry for creating so many issues lately!

EDIT: I don't think the problem is there, but have you ever considered replacing RSVP with bluebird for a probably faster promise implementation? Don't think that's going to happen anytime soon anyway.

**Comment by stefanpenner** [comment]
> but have you ever considered replacing RSVP with bluebird for a probably faster promise implementation?

they are actually [extremely close in performance](https://gist.github.com/stefanpenner/335ea79ad9e60d6cdc15) and in some scenarios RSVP is actually faster.

I am a-bit confused, I have provided context on how to resolve the issue (as requested), but it appears you are disregarding my suggestions. Which is frustrating, as you explicitly requested it but are disregarding it.

**Comment by stefanpenner** [comment]
> tree_merger_tmp_dest directories not getting cleaned up only on regular HDD mbp?

I believe this is actually the result of users doing `<ctrl-c>` skipping the cleanup. The concept of async finally is non-existent in node. We likely need to build a queue of work we must ensure has been pruned on process exist. In that handler, we will have to resort to sync (non-async) removal of contents.

**Comment by stefanpenner** [comment]
If you are legitimately interested in helping, are are able to commit to some reasonable amount of work. I can jump on a google hangout and brain dump as more required context as possible.

But only request that if you are serious about contributing.

**Comment by bucaran** [comment]
I am not disregarding anything, I am digesting everything. Let it settle in :smile: 

Thanks, I am definitely interested in helping, although I see ember-cli as a _problem_ right now because it's frustrating my team, I do find it's a complex piece of an even more complex puzzle and I've taken more than a handful of ideas that I hope to implement in Fly at some point. While I did find the lack of instrumentation in Funnel and Broc#builder annoying, at least it's instrumented at all!

I am super busy at work at the moment (plus it's 4 AM here in Tokyo), but if you are still available during the weekend let's hangout Sat or Sun, anytime :+1:

**Comment by stefanpenner** [comment]
> I am super busy at work at the moment (plus it's 4 AM here in Tokyo), but if you are still available during the weekend let's hangout Sat or Sun, anytime

Likely the following weekend will work, this next weekend I have family in town.

**Comment by bucaran** [comment]
No problem. I will make the time then for this and thanks.

**Comment by bucaran** [comment]
> I believe this is actually the result of users doing <ctrl-c> skipping the cleanup

BTW this is not the problem, I send the SIGINT only once and wait. I even added instrumentation to `.cleanup` after the promise is resolved and I get the rimraf events logged to the console.

```
broccoli-fun: ready to rimraf / /Users/b/dev/.../tmp/funnel-dest_LzJntx.tmp / +0ms
...
```

**Comment by stefanpenner** [comment]
It is possibly a timing issue, do to how slow HDD would be, manifesting as an early exit? Either before or during rimraf's thing.

**Comment by bucaran** [comment]
> It is possibly a timing issue, do to how slow HDD would be, manifesting as an early exit? Either before or during rimraf's thing. This seems strange, but \o/

Interesting. I will research how to prevent this and post here my findings. Thanks.

**Comment by stefanpenner** [comment]
> Interesting. I will research how to prevent this and post here my findings. Thanks.

Awesome, that would be handy.

**Comment by stefanpenner** [comment]
broccoli-filter is also known to leak its cache over time. https://github.com/caitp/cauliflower-filter/issues/12 mitigates this, and will benefit others once upstreamed into broccoli-filter (someone just needs to doit)

**Comment by bucaran** [comment]
Can we can get Broc's author @joliss to comment on this:

> Every stage in a broccoli pipeline is writing to disk, in an ember-cli build there are likely at least 20-30 stages that have to be gone through. That leads to massively different results between HDD and SDD (do to I/O throughput).

**Comment by rwjblue** [comment]
> Can we can get Broc's author @joliss to comment on this

For what purpose? What I said is correct, pinging everyone under the sun is unlikely to result in anything but annoying people that are working hard to solve these problems.

> How massive are these differences? (minutes?)

Feel free to test and report back.

> and have you investigated a viable solution?

Yes, we switched to using symlinks a while back which was a very large improvement, but writing anything to disk is slower (by quite a bit) on HDD vs SSD.  Also, the dynamic nature of the current Broccoli pipeline forces all plugins to be re-read for every rebuild (even if its input trees have not changed), that is being fixed with the refactor towards using [broccoli-plugin](https://github.com/broccolijs/broccoli-plugin) as a base class for all plugins.

**Comment by bucaran** [comment]
@rwjblue 
>  Also, the dynamic nature of the current Broccoli pipeline forces all plugins to be re-read for every rebuild (even if its input trees have not changed), that is being fixed with the refactor towards using broccoli-plugin as a base class for all plugins.

This is more or less what I was looking forward to get from this conversation. So, I'll apologize if I made you waste your time. On the other hand, this may be useful to other devs with the same problem.

One more thing:

> Broccoli pipeline forces all plugins to be re-read for every rebuild (even if its input trees have not changed),

What does it mean to _re-read_ a plugin? Do you mean that Broc will read, compile / transform and rewrite to disk on every file touch?

**Comment by stefanpenner** [comment]
> So, Broc will read, compile / transform and rewrite to disk on every file touch?

no, for more information please read http://www.solitr.com/blog/2014/02/broccoli-first-release/ which originates from https://github.com/broccolijs/broccoli#broccoli

**Comment by bucaran** [comment]
@stefanpenner I read that article before, I am aware of Broc basic technical details, but that was a reaction from what @rwjblue said.

> Broccoli pipeline forces all plugins to be re-read for every rebuild (even if its input trees have not changed),

Can you rephrase that? Because It seems my understanding was incorrect :disappointed_relieved:

**Comment by bucaran** [comment]
@rwjblue @stefanpenner 

Interesting discovery. I built my app and compared the state of the `tmp` directory in both the SSD and HDD mbps (other than that the configuration and project is exactly the same). The results were the HDD mpb is listing 379 directories and the SDD 700, almost twice as many directories.

What could the reason for this?

**Comment by bucaran** [comment]
Thanks, I have read the article and gained a better understanding of Broc's fundamental technical details :+1: 

What I said was in reaction to @rwjblue's:

> Broccoli pipeline forces all plugins to be re-read for every rebuild (even if its input trees have not changed),

It seems my understanding was incorrect :disappointed_relieved: , so can you rephrase that for me? 

And one more thing, digging into this could be useful to other folks out there.

So, I cleaned up the `tmp` directory in both machines, rebuilt, touch a file

```
touch app/app.coffee
```

and finally compared the state of the `tmp` directory in both the SSD and HDD mbps. Needless to say, all machines are running the same project and have matching configurations. I know the configurations are the same because I used the [same script](https://github.com/bucaran/rc/blob/master/bootstrap/main.sh) to setup the macs.


The results were the **SSD** lists way more directories and that they both have quite different directory structures.

So, while having a **SSD** certainly speeds things up, this data suggests `ember-cli` or `broccoli` is doing something quite different in both machines.

> I have removed the random strings from the result and sorted them both by repeated items:

**SSD MBP**  | HDD MBP
------------- | -------------
tree_merger-tmp_dest_dir (298)          | tree_merger-tmp_dest_dir (189)
reexporter-tmp_cache_dir (14)           | reexporter-tmp_cache_dir (10)
funnel-cache_path (105)                 | 
funnel-dest (79)                        | funnel-dest (105)
core_object-output_path (10)            | core_object-output_cache (10)
babel-cache_path (10)                   | babel-cache_path (10)
babel-output_path (9)                   |
coffee_script_filter-tmp_cache_dir (12) | coffee_script_filter-tmp_cache_dir (12)
core_object-cache_path (7)              | core_object-cache_path (8)
template_compiler-tmp_cache_dir (6)     | template_compiler-tmp_cache_dir (6)
caching-writer-dest-dir (10)            | caching-writer-dest-dir (6)
class-tmp_cache_dir (8)                 | class-tmp_cache_dir (4)
broccoli_merge_trees-cache_path (6)     |
broccoli_merge_trees-input_base_path (6)|
broccoli_merge_trees-output_path (6)    |
jshinter-tmp_cache_dir (4)              | jshinter-tmp_cache_dir (4)
                                        | read_compat-cache_path (4)
                                        | read_compat-input_base_path (4)
                                        | read_compat-output_path (4)
custom_replace-tmp_dest_dir (4)         | custom_replace-tmp_dest_dir (4)
core_object-dest_dir (4)                | core_object-dest_dir (2)
core_object-input_staging (4)           | core_object-input_staging (2)
sass_compiler-tmp_cache_dir             | sass_compiler-tmp_cache_dir (1)
static_compiler-tmp_dest_dir            | static_compiler-tmp_dest_dir (1)
config_loader-cache_path (3)            | config_loader-tmp_cache_dir (1)
core_object-output_cache (3)            | 



But why...? :sweat_smile:

**Comment by stefanpenner** [comment]
The article I linked to provides the information you ask for.

The diff in temp did is quite strange can you:

* share a simplified repo that demonstrates the same issue
* do a diff of the two node_module trees to confirm they are not different

**Comment by bucaran** [comment]
Thanks. It's going to be difficult or impossible to create a repo that can replicate the issue _unless_ the same occurs for a blank project, but I can definitely comply to the second request, I will get around that ASAP.

**Comment by stefanpenner** [comment]
> It's going to be difficult or impossible to create a repo that can replicate the issue unless the same occurs for a blank project,

it should be easy to test if a blank project, or a project with the same addons but very default app code.

**Comment by bucaran** [comment]
Yes, _that_ I can do with the help of a co-worker Monday. BTW the numbers in parentheses in the table above are the number of repetitions that directory was occurring inside the `tmp` directory. I removed the random part generated as that will obviously be different.

By the way, I instrumented `quicktemp` and found it to be one of the main bottlenecks in our application, not the only one, though.

**Comment by stefanpenner** [comment]
> By the way, I instrumented quicktemp and found it to be one of the main bottlenecks in our application, not the only one, though.

teardown or setup?

**Comment by bucaran** [comment]
Do you mean during the cleanup process? I saw a massive number of instances during cleanup, but it is during setup (startup) and (rebuilds) that the thing gets stuck (or rather getting a massive continuum of appearances. 

The others were `broc-writer` and `broc-filter` for sure.

**Comment by stefanpenner** [comment]
quicktemp on setup seems quite strange, it should just me making a single folder with some entropy for the name. Could we be starving for entropy at the part in time? Seems unlikely


If teardown is the problem, in theory each platform (FS) typically has an ultra performant way to purge large directory trees. Although the solutions may be not easy to bundle for all node users, it may be interesting to explore. A quick example: http://serverfault.com/questions/183821/rm-on-a-directory-with-millions-of-files/328305#328305

**Comment by bucaran** [comment]
@stefanpenner Thanks fro the ServerFault link, went straight to my bookmark list, a lot of so much good stuff in there.

The problem is definitely during the rebuilds, cleanup/teardown is done in no time.

**Comment by stefanpenner** [comment]
> The problem is definitely during the rebuilds, cleanup/teardown is done in no time.

Then why did you mention it.

> By the way, instrumenting quicktemp showed that was one of the main bottlenecks in our application. It's crazy :laughing:

**Comment by stefanpenner** [comment]
Ug, this conversation has returned to the points I raised in the original issue. Please only report new findings, that are very much validated. 

If you are looking to help, a good place to start is: https://github.com/ember-cli/ember-cli/issues/4562

**Comment by bucaran** [comment]
Wait, I didn't mention it, you asked me whether it was teardown or build and I said that they show up during teardown too, but it is the build/rebuild were the problem is.

**Comment by stefanpenner** [comment]
@bucaran 

> By the way, instrumenting quicktemp showed that was one of the main bottlenecks in our application. It's crazy :laughing:

**Comment by bucaran** [comment]
Well, I am not sure whether quicktemp is an issue, but before the instrumentation we were blacking out for 20 or 30 secs, and after it we found out most of these black holes were spent over at quicktemp summer house.

**Comment by stefanpenner** [comment]
https://github.com/broccolijs/broccoli-funnel/pull/29
